src/Sylius/Component/Addressing/spec/Matcher/ZoneMatcherSpec.php

```diff
@@ -13,274 +13,65 @@
+    function it_returns_all_matching_zones(
+        ZoneRepositoryInterface $zoneRepository,
         AddressInterface $address,
-        ZoneMemberInterface $memberProvince,
-        ZoneMemberInterface $memberCountry,
-        ZoneMemberInterface $memberZone,
-        ZoneInterface $zoneProvince,
-        ZoneInterface $zoneCountry,
-        ZoneInterface $zoneZone,
+        ZoneInterface $zoneOne,
+        ZoneInterface $zoneTwo,
+        ZoneInterface $zoneThree,
     ): void {
-        $repository->findAll()->willReturn([$zoneProvince, $zoneCountry, $zoneZone]);
-
-        $address->getProvinceCode()->willReturn('TX');
-        $memberProvince->getCode()->willReturn('TX');
-
-        $memberProvince->getBelongsTo()->willReturn($zoneProvince);
-        $zoneProvince->getType()->willReturn(ZoneInterface::TYPE_PROVINCE);
-        $zoneProvince->getMembers()->willReturn(new ArrayCollection([$memberProvince->getWrappedObject()]));
-
-        $address->getCountryCode()->willReturn('US');
-        $memberCountry->getCode()->willReturn('US');
-
-        $zoneCountry->getType()->willReturn(ZoneInterface::TYPE_COUNTRY);
-        $zoneCountry->getMembers()->willReturn(new ArrayCollection([$memberCountry->getWrappedObject()]));
-        $zoneCountry->getCode()->willReturn('USA');
-        $memberCountry->getBelongsTo()->willReturn($zoneCountry);
-
-        $memberZone->getCode()->willReturn('USA');
-        $zoneZone->getType()->willReturn(ZoneInterface::TYPE_ZONE);
-        $zoneZone->getMembers()->willReturn(new ArrayCollection([$memberZone->getWrappedObject()]));
-        $memberZone->getBelongsTo()->willReturn($zoneZone);
+        $zoneRepository->findAllByAddress($address)->willReturn([$zoneOne]);
+        $zoneRepository->findZonesByMembers([$zoneOne])->willReturn([$zoneTwo]);
+        $zoneRepository->findZonesByMembers([$zoneTwo])->willReturn([$zoneThree]);
+        $zoneRepository->findZonesByMembers([$zoneThree])->willReturn([]);
 
-        $repository->findOneBy(['code' => 'USA'])->willReturn($zoneCountry);
+        $matchedZones = $this->matchAll($address);
 
-        $this->matchAll($address)->shouldReturn([$zoneProvince, $zoneCountry, $zoneZone]);
+        $matchedZones->shouldHaveCount(3);
+        $matchedZones->shouldBe([$zoneOne, $zoneTwo, $zoneThree]);
     }
 
-    function it_matches_all_zones_by_scope_when_one_zone_for_address_is_defined(
-        RepositoryInterface $repository,
+    function it_returns_all_matching_zones_withing_a_matching_scope(
+        ZoneRepositoryInterface $zoneRepository,
         AddressInterface $address,
-        ZoneMemberInterface $memberCountry,
-        ZoneInterface $zoneCountry,
+        ZoneInterface $zoneOne,
+        ZoneInterface $zoneTwo,
+        ZoneInterface $zoneThree,
     ): void {
-        $repository->findBy(['scope' => ['shipping', 'all']])->willReturn([$zoneCountry]);
+        $zoneOne->getScope()->willReturn('nato');
+        $zoneTwo->getScope()->willReturn('all');
```

Comment:
I'm not suggesting this change, but noting, that it could be done.

```patch
--- $zoneTwo->getScope()->willReturn('all');
+++ $zoneTwo->getScope()->willReturn(Scope::ALL);
```